[CI] TermsEnumTests testTermsEnumIPRandomized failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #11890 / rocky-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/typo65nn27nii)
- [elasticsearch-periodic-platform-support #11890 / rocky-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/uhrb3okyv3wh4)
- [elasticsearch-pull-request #122282 / part-2](https://gradle-enterprise.elastic.co/s/fumuw7n4pu3do)
- [elasticsearch-pull-request #122274 / part-2](https://gradle-enterprise.elastic.co/s/wfxxp4zroxuro)
- [elasticsearch-pull-request #121874 / part-2](https://gradle-enterprise.elastic.co/s/ugs27jbcjnrsk)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:core:test" --tests "org.elasticsearch.xpack.core.termsenum.TermsEnumTests.testTermsEnumIPRandomized" -Dtests.seed=4ED6E49608803825 -Dtests.locale=en-NL -Dtests.timezone=America/Vancouver -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.core.termsenum.TermsEnumTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testTermsEnumIPRandomized'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected 1 for prefix 0 but was 0, [] expected:<1> but was:<0>
```

**Issue Reasons:**
- [main] 2 consecutive failures in step rocky-8_platform-support-unix
- [main] 5 failures in test testTermsEnumIPRandomized (1.2% fail rate in 428 executions)
- [main] 2 failures in step rocky-8_platform-support-unix (50.0% fail rate in 4 executions)
- [main] 3 failures in step part-2 (1.1% fail rate in 274 executions)
- [main] 3 failures in pipeline elasticsearch-pull-request (1.1% fail rate in 269 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/docs/changelog/144308.yaml
+++ b/docs/changelog/144308.yaml
@@ -2,5 +2,5 @@ area: Search
 issues:
  - 142811
 pr: 144308
-summary: Fix edge case in "_terms_enum" API for ip fields
+summary: Fixes an edge case in "_terms_enum" API for ip fields
 type: bug

--- a/docs/changelog/144308.yaml
+++ b/docs/changelog/144308.yaml
@@ -1,0 +1,6 @@
+area: Search
+issues:
+ - 142811
+pr: 144308
+summary: Fix edge case in "_terms_enum" API for ip fields
+type: bug

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -249,9 +249,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.tree.EsqlNodeSubclassTests
   method: testTransform {class org.elasticsearch.xpack.esql.plan.logical.MMR}
   issue: https://github.com/elastic/elasticsearch/issues/142674
-- class: org.elasticsearch.xpack.core.termsenum.TermsEnumTests
-  method: testTermsEnumIPRandomized
-  issue: https://github.com/elastic/elasticsearch/issues/142811
 - class: org.elasticsearch.reindex.management.ReindexManagementClientYamlTestSuiteIT
   method: test {yaml=reindex/30_cancel_reindex/Cancel running reindex returns response and GET confirms completed}
   issue: https://github.com/elastic/elasticsearch/issues/142079

--- a/server/src/main/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtil.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtil.java
@@ -34,6 +34,13 @@ public class IpPrefixAutomatonUtil {
 
     private static final Automaton EMPTY_AUTOMATON = Automata.makeEmpty();
     private static final Automaton IPV4_PREFIX = Automata.makeBinary(new BytesRef(new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, -1 }));
+    private static final Automaton ZERO_FIRST_GROUP = concatenate(List.of(Automata.makeChar(0), Automata.makeChar(0)));
+    private static final Automaton NON_ZERO_SECOND_GROUP = Operations.union(
+        List.of(
+            concatenate(List.of(Automata.makeCharRange(1, 255), Automata.makeCharRange(0, 255))),
+            concatenate(List.of(Automata.makeChar(0), Automata.makeCharRange(1, 255)))
+        )
+    );
 
     static final Map<Integer, Automaton> INCOMPLETE_IP4_GROUP_AUTOMATON_LOOKUP = new HashMap<>();
     static {
@@ -107,15 +114,22 @@ public class IpPrefixAutomatonUtil {
                 } else {
                     // potentially partial block
                     if (groupsAdded == 0 && ONLY_ZEROS.matcher(group).matches()) {
-                        // here we have a leading group with only "0" characters. If we allowed this to match
-                        // ipv6 addresses, this would include things like 0000::127.0.0.1 (and all other ipv4 addresses).
-                        // Allowing this would be counterintuitive, so "0*" prefixes should only expand
-                        // to ipv4 addresses like "0.1.2.3" and we return with an automaton not matching anything here
-                        return EMPTY_AUTOMATON;
+                        if (group.length() == 1) {
+                            // A single "0" can match IPv6 addresses displayed as "0:X:..." where the first group is
+                            // exactly 0x0000. We require the second group to be non-zero to avoid also matching all
+                            // IPv4-mapped addresses (::ffff:x.x.x.x) and addresses where "::" compresses leading zeros.
+                            ipv6Automaton = concatenate(List.of(ipv6Automaton, ZERO_FIRST_GROUP, NON_ZERO_SECOND_GROUP));
+                            groupsAdded = 2;
+                        } else {
+                            // Multi-zero prefixes ("00", "000", "0000") cannot match any displayed IPv6 group
+                            // since leading zeros are always stripped in IPv6 formatting.
+                            return EMPTY_AUTOMATON;
+                        }
+                    } else {
+                        // we need to create all possibilities of byte sequences this could match
+                        groupsAdded++;
+                        ipv6Automaton = concatenate(ipv6Automaton, automatonFromIPv6Group(group));
                     }
-                    // we need to create all possibilities of byte sequences this could match
-                    groupsAdded++;
-                    ipv6Automaton = concatenate(ipv6Automaton, automatonFromIPv6Group(group));
                 }
             }
             // fill up the remainder of the 16 address bytes with wildcard matches, each group added so far counts for two bytes

--- a/server/src/test/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtilTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtilTests.java
@@ -113,7 +113,7 @@ public class IpPrefixAutomatonUtilTests extends ESTestCase {
             assertEquals("::7f00:1", NetworkAddress.format(getByName("0000::127.0.0.1")));
             // same for more than one leading 0, those get condensed to "::" in the terms API output, so we shouldn't match them
             assertFalse(accepts(a, "0:0:0::127.0.0.1"));
-            assertEquals("::7f00:1", NetworkAddress.format(getByName("0000::127.0.0.1")));
+            assertEquals("::7f00:1", NetworkAddress.format(getByName("0:0:0::127.0.0.1")));
             assertFalse(accepts(a, "::1"));
             assertFalse(accepts(a, "::"));
             assertFalse(accepts(a, "::1:0:0:0:0:0"));

--- a/server/src/test/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtilTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/IpPrefixAutomatonUtilTests.java
@@ -21,6 +21,7 @@ import java.net.InetAddress;
 import java.net.UnknownHostException;
 import java.util.Arrays;
 
+import static java.net.InetAddress.getByName;
 import static org.elasticsearch.index.mapper.IpPrefixAutomatonUtil.buildIpPrefixAutomaton;
 import static org.elasticsearch.index.mapper.IpPrefixAutomatonUtil.parseIp6Prefix;
 import static org.hamcrest.Matchers.contains;
@@ -101,15 +102,29 @@ public class IpPrefixAutomatonUtilTests extends ESTestCase {
         {
             CompiledAutomaton a = buildIpPrefixAutomaton("0");
             assertTrue(accepts(a, "0.2.3.4"));
+            assertTrue(accepts(a, "0:1234::1"));
+            assertTrue(accepts(a, "0:abcd:ef01:2345:6789:abcd:ef01:2345"));
+            assertTrue(accepts(a, "0:517::b12a:6a52:0:0"));
             assertFalse(accepts(a, "::1.1.2.3"));
             assertFalse(accepts(a, "1c7:21d6:ffff:ffff:6852:f279::"));
+            // the next is not an IP4 address, so the condensed display form we are expecting from the TermsEnum
+            // output is "::7f00:1" and shouldn't match
             assertFalse(accepts(a, "0000::127.0.0.1"));
+            assertEquals("::7f00:1", NetworkAddress.format(getByName("0000::127.0.0.1")));
+            // same for more than one leading 0, those get condensed to "::" in the terms API output, so we shouldn't match them
+            assertFalse(accepts(a, "0:0:0::127.0.0.1"));
+            assertEquals("::7f00:1", NetworkAddress.format(getByName("0000::127.0.0.1")));
+            assertFalse(accepts(a, "::1"));
+            assertFalse(accepts(a, "::"));
+            assertFalse(accepts(a, "::1:0:0:0:0:0"));
         }
         {
             CompiledAutomaton a = buildIpPrefixAutomaton("00");
             assertFalse(accepts(a, "0.2.3.4"));
             assertFalse(accepts(a, "::1.1.2.3"));
             assertFalse(accepts(a, "c7:21d6:ffff:ffff:6852:f279::"));
+            assertFalse(accepts(a, "0:1234::1"));
+            assertFalse(accepts(a, "0:abcd:ef01:2345:6789:abcd:ef01:2345"));
             assertFalse(accepts(a, "0000::127.0.0.1"));
         }
         {
@@ -117,6 +132,8 @@ public class IpPrefixAutomatonUtilTests extends ESTestCase {
             assertTrue(accepts(a, "0.2.3.4"));
             assertTrue(accepts(a, "::1.1.2.3"));
             assertTrue(accepts(a, "0:21d6:ffff:ffff:6852:f279::"));
+            assertTrue(accepts(a, "0:1234::1"));
+            assertTrue(accepts(a, "0:abcd:ef01:2345:6789:abcd:ef01:2345"));
             assertTrue(accepts(a, "0000::127.0.0.1"));
             assertFalse(accepts(a, "0001:21d6:ffff:ffff:6852:f279::"));
         }
@@ -183,7 +200,7 @@ public class IpPrefixAutomatonUtilTests extends ESTestCase {
     }
 
     private static boolean accepts(CompiledAutomaton compiledAutomaton, String address) throws UnknownHostException {
-        byte[] encoded = InetAddressPoint.encode(InetAddress.getByName(address));
+        byte[] encoded = InetAddressPoint.encode(getByName(address));
         return compiledAutomaton.runAutomaton.run(encoded, 0, encoded.length);
     }
 

